### PR TITLE
Byte decode error

### LINKED ISSUE #88: Support custom encoding/decoding errors
It would be better and easier to debug a program if the `BytesEncoding` and `BytesDecoding` traits could return any error type.

To do so we need to modify the `Error` enum and more specifically the `Encoding` and `Decoding` variants to wrap a `Box<dyn Error>`.

### PR DESCRIPTION
**Pull request by MarinPostma**
close #88

### CODE REVIEW

File: heed-types/src/cow_slice.rs [hunk 1]
```diff
@@ -58,10 +59,10 @@ where
                         vec.set_len(elems);
                     }
 
-                    return Some(Cow::Owned(vec));
+                    return Ok(Cow::Owned(vec));
                 }
 
-                None
+                Err("The provided bytes do not satisfy the alignment requirements.")?
```
Comment (Kerollmops) [inline heed-types/src/cow_slice.rs:65]: When I switch from zerocopy to bytemuck I will change this error string into [a nicer error, more precise](https://docs.rs/bytemuck/1.4.1/bytemuck/enum.PodCastError.html)!

File: heed-types/src/lib.rs [hunk 2]
```diff
@@ -71,8 +71,8 @@ pub struct DecodeIgnore;
 impl heed_traits::BytesDecode<'_> for DecodeIgnore {
     type DItem = ();
 
-    fn bytes_decode(_bytes: &[u8]) -> Option<Self::DItem> {
-        Some(())
+    fn bytes_decode(_bytes: &[u8]) -> Result<Self::DItem, Box<dyn std::error::Error>> {
```
Comment (Kerollmops) [inline heed-types/src/lib.rs:74]:
```suggestion
    fn bytes_decode(_bytes: &[u8]) -> Result<Self::DItem, Box<dyn Error>> {
```

By importing the Error trait into the scope.

File: heed/src/db/polymorph.rs [hunk 3]
```diff
@@ -278,7 +278,7 @@ impl PolyDatabase {
     {
         assert_eq!(self.env_ident, txn.env.env_mut_ptr() as usize);
 
-        let key_bytes: Cow<[u8]> = KC::bytes_encode(&key).ok_or(Error::Encoding)?;
+        let key_bytes: Cow<[u8]> = KC::bytes_encode(&key).map_err(|e| Error::Encoding(e))?;
```
Comment (Kerollmops) [inline heed/src/db/polymorph.rs:281]:
```suggestion
        let key_bytes: Cow<[u8]> = KC::bytes_encode(&key).map_err(Error::Encoding)?;
```

File: heed/src/lib.rs [hunk 4]
```diff
@@ -80,8 +80,8 @@ use std::{error, fmt, io, result};
 pub enum Error {
     Io(io::Error),
     Mdb(MdbError),
-    Encoding,
-    Decoding,
+    Encoding(Box<dyn std::error::Error>),
+    Decoding(Box<dyn std::error::Error>),
```
Comment (Kerollmops) [inline heed/src/lib.rs:84]:
```suggestion
    Encoding(Box<dyn Error>),
    Decoding(Box<dyn Error>),
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]
Thank you, I tried but haven't finished yet. So, you take the lead and I will accept your PR!

----

I am wondering if introducing a custom Result type named `CodecResult`, could be better.
```
type CodecResult<T> = Result<T, Box<dyn Error>>;
```

**Comment by MarinPostma** [comment]
> I am wondering if introducing a custom public Result type named CodecResult, could be better.

Why not exposing an anyhow Result instead?

**Comment by Kerollmops** [comment]
> Why not exposing an anyhow Result instead?

I don't want to bring one more dependency in this crate, don't want to force users to use anyhow, and anyhow has been design to be used by binary crates.

Exposing a type alias will work just fine :)

**Pull request closed without merging**